gae_dashboard/waf_alert.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
+"""
+Periodically checks blocked requests which indicates either
+an attack or a problem with the WAF rules.
+"""
+
+import datetime
+import alertlib
+import bq_util
+import argparse
+
+BQ_PROJECT = "khanacademy.org:deductive-jet-827"
+FASTLY_DATASET = "fastly"
+FASTLY_WAF_LOG_TABLE_PREFIX = "khanacademy_dot_org_waf_logs"
+FASTLY_LOG_TABLE_PREFIX = "khanacademy_dot_org_logs"
+
+# The size of the period of time to query. We are monitoring every 60 minutes.
+WAF_PERIOD = 60 * 60
+
+# Spike percentage determined from WAF Investigation:
+# https://app.mode.com/editor/khanacademy/reports/f58d1ec83e48/presentation
+SPIKE_PERCENTAGE = 0.03
+
+TABLE_FORMAT = "%Y%m%d"
+TS_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+ALERT_CHANNEL = "#bot-testing"
+
+BLOCKED_REQ = """
+#standardSQL
+SELECT
+  COUNT(*) AS blocked,
+  APPROX_TOP_COUNT(client_ip,1)[OFFSET(0)].value AS ip,
+  100 * APPROX_TOP_COUNT(client_ip,1)[OFFSET(0)].count /
+   COUNT(*) AS ip_percentage,
+  APPROX_TOP_COUNT(waf.message,1)[OFFSET(0)].value AS message,
+  100 * APPROX_TOP_COUNT(waf.message,1)[OFFSET(0)].count /
+   COUNT(*) AS message_percentage,
+  APPROX_TOP_COUNT(request_user_agent,1)[OFFSET(0)].value
+   AS request_user_agent,
+  100 * APPROX_TOP_COUNT(request_user_agent,1)[OFFSET(0)].count
+   / COUNT(*) AS request_user_agent_percentage,
+FROM
+    `{fastly_waf_log_tables}`
+  WHERE
+    TIMESTAMP(timestamp) BETWEEN TIMESTAMP('{start_timestamp}')
+    AND TIMESTAMP('{end_timestamp}')
+    AND blocked=1
+    AND _TABLE_SUFFIX BETWEEN '{yesterday_ymd}' AND '{ymd}'
+"""
+
+TOTAL_REQ = """
+#standardsql
+SELECT
+  count(*) AS total
+FROM
+  `{fastly_log_tables}`
+WHERE
+  TIMESTAMP(timestamp) BETWEEN TIMESTAMP('{start_timestamp}')
+  AND TIMESTAMP('{end_timestamp}')
+  AND _TABLE_SUFFIX BETWEEN '{yesterday_ymd}' AND '{ymd}'
+"""
+
+
+# Detects a WAF attack and alerts to Slack channel if there is an attack.
+def waf_detect(endtime):
+    ymd = endtime.strftime(TABLE_FORMAT)
+
+    # For edge case where we need today and yesterday's log (23:00:00 - 00:00:00)
```

Comment:
This looks correct!  I think this can be simplified as:

```suggestion
    start_ymd = start.strftime(TABLE_FORMAT)
```

Then we don't need to craft out the if clause?